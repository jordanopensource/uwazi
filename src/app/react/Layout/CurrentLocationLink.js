import React from 'react';
import PropTypes from 'prop-types';
import { Link, useLocation } from 'react-router-dom';

const validProps = props => {
  const { to, ...valid } = props;
  return valid;
};

const CurrentLocationLink = ({ children, queryParams, ...otherProps }) => {
  const location = useLocation();
  const query = new URLSearchParams(location.search);
  Object.keys(queryParams).forEach(key => {
    query.set(key, queryParams[key]);
  });

  query.forEach((value, key) => {
    if (value === '') {
      query.delete(key);
    }
  });

  return (
    <Link to={`${location.pathname}?${query.toString()}`} {...validProps(otherProps)}>
      {children}
    </Link>
  );
};

CurrentLocationLink.defaultProps = {
  children: '',
  queryParams: {},
};

CurrentLocationLink.propTypes = {
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  queryParams: PropTypes.object, // eslint-disable-line react/forbid-prop-types
};

export { CurrentLocationLink };

export default CurrentLocationLink;
