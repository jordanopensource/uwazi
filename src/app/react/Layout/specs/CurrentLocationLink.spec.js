import React from 'react';
import { shallow } from 'enzyme';
import { CurrentLocationLink } from '../CurrentLocationLink';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({ pathname: 'pathanem', search: 'param=value&param2=value2' }),
}));

describe('Link', () => {
  let component;
  const props = { queryParams: {}, to: '', className: '', prop: '' };

  const render = () => {
    component = shallow(<CurrentLocationLink {...props}>text</CurrentLocationLink>);
  };

  it('should render a link with current location as the url', () => {
    render();
    expect(component).toMatchSnapshot();
  });

  it('should pass all other props to Link', () => {
    props.queryParams = { param2: '', test: false };
    props.to = 'not to be passed';
    props.className = 'className';
    props.prop = 'value';

    render();
    expect(component).toMatchSnapshot();
  });

  it('should remove props when they are an empty string', () => {
    props.queryParams = { param2: '', test: false };
    render();
    expect(component).toMatchSnapshot();
  });

  it('should overwrite url query with query params passed (without mutating query)', () => {
    props.queryParams = { param2: 'new value', test: 'test' };
    render();
    expect(component).toMatchSnapshot();
    expect(component.props().to).toEqual('pathanem?param=value&param2=new+value&test=test');
  });
});
