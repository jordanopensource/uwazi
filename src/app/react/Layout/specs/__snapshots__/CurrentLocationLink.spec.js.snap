// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should overwrite url query with query params passed (without mutating query) 1`] = `
<Link
  className="className"
  prop="value"
  to="pathanem?param=value&param2=new+value&test=test"
>
  text
</Link>
`;

exports[`Link should pass all other props to Link 1`] = `
<Link
  className="className"
  prop="value"
  to="pathanem?param=value&test=false"
>
  text
</Link>
`;

exports[`Link should remove props when they are an empty string 1`] = `
<Link
  className="className"
  prop="value"
  to="pathanem?param=value&test=false"
>
  text
</Link>
`;

exports[`Link should render a link with current location as the url 1`] = `
<Link
  className=""
  prop=""
  to="pathanem?param=value&param2=value2"
>
  text
</Link>
`;
